src/controllers/bulk/bulksupported.h

```diff
@@ -1,18 +1,28 @@
+#pragma once
+
+#include <cstdint>
+
+struct bulk_device_id {
+    std::uint16_t vendor_id;
+    std::uint16_t product_id;
+};
+
 // A list of supported USB bulk devices
 
-#pragma once
+struct bulk_device_tweak {
+    std::uint8_t in_epaddr;
+    std::uint8_t out_epaddr;
+    std::uint8_t interface_number;
+};
 
-typedef struct bulk_supported {
-    unsigned short vendor_id;
-    unsigned short product_id;
-    unsigned char in_epaddr;
-    unsigned char out_epaddr;
-    unsigned int interface_number;
-} bulk_supported_t;
+struct bulk_support_lookup {
+    bulk_device_id key;
+    bulk_device_tweak value;
+};
 
-static bulk_supported_t bulk_supported[] = {
-        {0x06f8, 0xb105, 0x82, 0x03, 0}, // Hercules MP3e2
-        {0x06f8, 0xb107, 0x83, 0x03, 0}, // Hercules Mk4
-        {0x06f8, 0xb100, 0x86, 0x06, 0}, // Hercules Mk2
-        {0x06f8, 0xb120, 0x82, 0x03, 0}, // Hercules MP3 LE / Glow
-        {0, 0, 0, 0, 0}};
+constexpr static bulk_support_lookup bulk_supported[] = {
+        {{0x06f8, 0xb105}, {0x82, 0x03, 0}}, // Hercules MP3e2
+        {{0x06f8, 0xb107}, {0x83, 0x03, 0}}, // Hercules Mk4
```

Comment:
no idea. @JoergAtGithub may know.

Reply:
Workaround: Add a TODO, and use -1 as interfac number.

Reply:
> Interestingly the MK4 is reported here at: https://mixxx.discourse.group/t/hercules-mk4-hid-under-linux-and-1-11/13447/3 {0x83, 0x02, 0} and {0x83, 0x03, 5} Can we have the endpoint address twice? Is that the same end point?

This is not the same USB endpoint, because only the address `0x83` is identical, but not the number of the endpoint (`0x02` vs. `0x03`).

Endpoints and Interfaces are fields on different USB protocol layers:

- Endpoints are raw hardware addresses which the operating systems USB kernel driver uses to prioritize the data transports with guranteed latency (Isochronous or Interupt transport mode) and non real-time transports (Control or Bulk transport mode)
- Interfaces/Pipes are the a protocol level higher and depend not only on the hardware, but also on configuration from the host software (drivers like power-management or user mode applications). You can assign a Endpoint only to one USB Interface, otherwise you could get conflicting access. But an interface can run in different alternative modes (e.g. a camera can run in realtime mode for video using the endpoint Isocronous transfer mode or in high-resolution photo mode using the same endpoint in Bulk transfer mode instead).

See: https://beyondlogic.org/usbnutshell/usb5.shtml

Reply:
I am still confused, because we don't use endpoint number in our code only the address.
Just double checked but I don't see the double 0x83 anymore, instead I found a double 0x81 with: 
Interface 1 Endpoint Address 0x81
Interface 3 Endpoint Address 0x81
Interface 4 Endpoint Address 0x81

Reply:
This information is not unambiguous. If you talk about the transport layer need to specify the pair of Endpoint Address and Enpoint Number: 
![endpoint](https://github.com/user-attachments/assets/d55d5b97-f4d5-46ef-90d8-9866afce8b3f)

Reply:
It is still puzzling form me, because we do not consider the Interface number with Linux at all but it seams to work with the enpoint adress only, even without an Endpoint number.  libusb_claim_interface() is written in a way that we may claim multiple interfaces with one handel. 
https://github.com/mixxxdj/mixxx/blob/4df25086d1aa060dcc3b2793bc5be20c7380e0a1/src/controllers/bulk/bulkcontroller.cpp#L199
libusb_bulk_transfer

Here we use only the enpoint adress 
https://github.com/mixxxdj/mixxx/blob/4df25086d1aa060dcc3b2793bc5be20c7380e0a1/src/controllers/bulk/bulkcontroller.cpp#L42C18-L42C38

The underlying questions is what is the interface number for -1 and will it work like before? 
How does the driver know which endpoint shall be used if there are tw owith the same address. 
Or are they the same listed in two interfaces?

Reply:
> I am still confused, because we don't use endpoint number in our code only the address.
> Just double checked but I don't see the double 0x83 anymore, instead I found a double 0x81 with:
> Interface 1 Endpoint Address 0x81
> Interface 3 Endpoint Address 0x81
> Interface 4 Endpoint Address 0x81

I double checked this, and I found out, that the Endpoint-Number is part of the Endpoint-Address:
Each Interfaces can have up to 16 Endpoints and the address is a single byte build as follows:

- Bits 0..3b Endpoint Number.
- Bits 4..6b Reserved. Set to Zero
- Bits 7 Direction 0 = Out, 1 = In (Ignored for Control Endpoints)

![intftree](https://github.com/user-attachments/assets/50d9792c-4e44-4c1e-9f49-a99189b4290a)

See https://beyondlogic.org/usbnutshell/usb5.shtml for reference